master/buildbot/util/kubeclientservice.py

```diff
@@ -190,19 +193,31 @@ def __init__(self, kube_config=None):
         self.config = kube_config
         HTTPClientService.__init__(self, '')
         self._namespace = None
-        self.addService(kube_config)
+        kube_config.setServiceParent(self)
 
+    @defer.inlineCallbacks
     def _prepareRequest(self, ep, kwargs):
         config = self.config.getConfig()
         self._base_url = config['master_url']
         url, req_kwargs = HTTPClientService._prepareRequest(self, ep, kwargs)
 
+        if config['headers']:
+            if not req_kwargs['headers']:
+                req_kwargs['headers'] = {}
+            else:
+                req_kwargs['headers'] = req_kwargs['headers'].dup
+            req_kwargs['headers'].update(config['headers'])
+
+        auth = config.get('authorization', None)
+        if auth is not None:
+            req_kwargs['headers']['Authorization'] = yield self.config.renderSecrets(auth)
```

Comment:
renderSecrets can be slow (contact Vault for example). generally, the secrets are cached. It would be better if the secrets were calculated at the time of the reconfigService, so that we don't have to put prepareRequest as returning deferred.

Reply:
As far as I can see, that would require a major infrastructure change in buildbot - reconfigService is not expected to return a deferred. I'm not ready to spend the time needed for that change given that "slow" is not really an issue here - starting the builds on GKE for me take already ~10s, I don't mind having a few more.

Reply:
yes, reconfigService is optionally deferable. checkConfig is not.

The only thing is to make sure secretManager is available at that time. The other option is to just cache those secrets. I agree this can be done in a later PR if you encounter some performance issues